cpp/oneapi/dal/algo/kmeans/backend/gpu/kernels_csr_impl.hpp

```diff
@@ -49,91 +52,83 @@ sycl::event compute_data_squares(sycl::queue& q,
-    const auto row_ofs = row_offsets.get_data();
-    const auto col_ind = column_indices.get_data();
-    const auto b_ptr = b.get_data();
-
-    // Compute matrix block by block to avoid integer overflow
-    const std::int64_t row_block = 8 * bk::device_max_wg_size(q);
-    const std::int64_t row_block_size = std::min<std::int64_t>(row_block, a_row_count);
-    const std::int64_t col_block_size = std::min<std::int64_t>(row_block, b_row_count);
-
-    const auto nd_range =
-        bk::make_multiple_nd_range_3d({ row_block_size, col_block_size, local_size },
-                                      { 1, 1, local_size });
-
-    return q.submit([&](sycl::handler& cgh) {
+sycl::event transpose(sycl::queue& q,
+                      const pr::ndview<Float, 2>& src,
+                      pr::ndview<Float, 2>& dst,
+                      const event_vector& deps = {}) {
+    const auto src_shape = src.get_shape();
+    ONEDAL_ASSERT(src_shape[0] > 0);
+    ONEDAL_ASSERT(src_shape[1] > 0);
+    ONEDAL_ASSERT(src_shape[0] == dst.get_dimension(1));
+    ONEDAL_ASSERT(src_shape[1] == dst.get_dimension(0));
+    const auto row_count = src_shape[0];
+    const auto col_count = src_shape[1];
+
+    const auto range = sycl::range<2>(row_count, col_count);
+
+    const Float* src_ptr = src.get_data();
+    Float* dst_ptr = dst.get_mutable_data();
+    auto event = q.submit([&](sycl::handler& cgh) {
         cgh.depends_on(deps);
-        cgh.parallel_for(nd_range, [=](auto item) {
-            const auto row_shift = item.get_global_id(0);
-            const auto col_shift = item.get_global_id(1);
-            const auto local_id = item.get_local_id(2);
-
-            for (auto row_idx = row_shift; row_idx < a_row_count; row_idx += row_block) {
-                for (auto col_idx = col_shift; col_idx < b_row_count; col_idx += row_block) {
-                    const auto start = row_ofs[row_idx] + local_id;
-                    const auto end = row_ofs[row_idx + 1];
-                    Float acc = Float(0);
-                    for (std::int64_t data_idx = start; data_idx < end; data_idx += local_size) {
-                        const auto reduce_id = col_ind[data_idx];
-                        acc += a_ptr[data_idx] * b_ptr[col_idx * reduce_dim + reduce_id];
-                    }
-                    const Float scalar_mul =
-                        sycl::reduce_over_group(item.get_group(),
-                                                acc,
-                                                Float(0),
-                                                sycl::ext::oneapi::plus<Float>());
-                    if (local_id == 0) {
-                        res_ptr[row_idx * b_row_count + col_idx] =
-                            beta * res_ptr[row_idx * b_row_count + col_idx] + alpha * scalar_mul;
-                    }
-                }
-            }
+        cgh.parallel_for(range, [=](auto item) {
+            auto i = item[0];
```

Comment:
const std::int64_t i = item[0], j = item[1]; is probably better

Reply:
Done.